modin/pandas/groupby.py

```diff
@@ -294,6 +296,34 @@ def first(self, **kwargs):
     def backfill(self, limit=None):
         return self.bfill(limit)
 
+    _internal_by_cache = no_default
+
+    # TODO: since python 3.9:
+    # @cached_property
+    @property
+    def _internal_by(self):
+        """
+        Get only those components of 'by' that are column labels of the source frame.
+
+        Returns
+        -------
+        list of labels
+        """
+        if self._internal_by_cache is not no_default:
+            return self._internal_by_cache
+
+        internal_by = []
```

Comment:
can we use a tuple here (as this should be a read-only thing, plus tuples eat less memory)?

Reply:
We can't do it efficiently, since it's required to do both: appending a single item and extending the container with a sequence, when filling `internal_by` (so no list comprehension here). 

Treating tuple in the same way as a list in a pure for-loops hurts performance, since tuple is immutable, so no "amortized O(1)" insertions here:

<details><summary><i>"Microbenchmarks to the rescue!"</i></summary>

```python
import timeit

N = 1000
M = 10

def list_append(): # timeit() -> 0.41s
    container = list()
    for i in range(N):
        if i % 10:
            container.extend(range(M))
        else:
            container.append(i)

def tuple_append(): # timeit() -> 12.24s
    container = tuple()
    for i in range(N):
        if i % 10:
            container += tuple(range(M))
        else:
            container += (i,)


t1 = timeit.timeit(list_append, number=1000)
t2 = timeit.timeit(tuple_append, number=1000)

print(t1)
print(t2)
```

</details>

We can cast just cast list to a tuple at the end of the function, but will it make sense then?

Reply:
rewritten this function so it uses list-comprehension here (and so using a tuple)